### PR TITLE
Add support join hits lookup by id

### LINKED ISSUE #76133: Support field retrieval joins
Today, the [enrich processor ](https://www.elastic.co/guide/en/elasticsearch/reference/current/enrich-processor.html) supports enriching data seamlessly at indexing time. But users must reindex whenever the enrich index changes. This issue proposes to support enriching field retrieval at query time as an alternative. The proposal only supports field retrieval joins; searches and aggregation aren't covered. Like [runtime fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/runtime.html), field retrieval joins also reduce disk usage and indexing time because enriched fields are not looked up and indexed. However, field retrieval joins are not suitable if they are retrieved often or searched or aggregated.

Below is an example of the proposed API. It supports look up by the` _id` of the target index.

```js
PUT /users/_doc/mardy.brown?refresh=wait_for
{
  "email": "mardy.brown@asciidocsmith.com",
  "first_name": "Mardy",
  "last_name": "Brown",
  "city": "New Orleans",
  "county": "Orleans",
  "state": "LA",
  "zip": 70116,
  "web": "mardy.asciidocsmith.com",
  "joined_date": "2021-01-02"
}
```

```js
POST /blogs/_doc?refresh=wait_for
{
  "title": "Elasticsearch is fast",
  "author": "mardy.brown"
}
```

```js
POST /blogs/_search
{
  "query": {
    "match": {
      "title": "fast"
    }
  },
  "fields": [
    "title",
    {
      "field": "author.first_name",
      "lookup_index": "users",
      "match_field": "author",
      "lookup_path": "first_name" 
    },
    {
      "field": "author.last_name", // optional, if not specified then it's `{{lookup_index}}.{{lookup_path}}`
      "lookup_index": "users",
      "match_field": "author",
      "lookup_path": "last_name"
    },
    {
      "field": "author.joined_date",
      "lookup_index": "users",
      "match_field": "author",
      "format": "yyyy-MM-dd",
      "lookup_path": "joined_date"
    }
  ]
}
```

Should we also support queries in the lookup? I can be useful, but I don't find a good use case.

```js
POST /blogs/_search
{
  "query": {
    "match": {
      "title": "cool"
    }
  },
  "fields": [
    "title",
    {
      "field": "first_name",
      "lookup_index": "users",
      "query" : {
          "match": {
              "email": "{{email}}"
          }
      }
    }
  ]
}
```

Should we also support field retrieval from the aggregated result?

```js
GET restaurant/_search
{
  "query": {
    "match": {
      "menu": "sushi"
    }
  },
  "fields": [
    "name",
    {
      "field": "rating",
      "lookup_index": "reviews",
      "query": {
        "term": {
          "_id": "{{restaurant_id}}"
        }
      },
      "aggs": {
        "avg_rating":{"avg": { "field": "rating" }}
      },
      "aggs_path": "aggregations.avg_rating.value"
    }
  ]
}
```

A field retrieval joins can be added to the mapping of an index.

Feedback on the API and these questions are welcomed.

### PR DESCRIPTION
**Pull request by dnhatn**
The new option `join_hits` is used to enrich search hits with associated documents from related indices.

The following search returns the detailed document from the `sensors` index for each search hit from the `sensor-logs` index.

```js
GET sensor-logs/_search
{
 "query": {
  "range": {
    "temperature": {
      "gt": 100
    }
  }
 }
 "join_hits": [
  {
    "name": "sensor",
    "index": "sensors",
    "match_field": "sensor_id"
  }
 ]
}
```

The search returns

```js
{
  ...,
  "hits":{
    ...,
    "hits":[
      {
        "_id": "hit_id_0",
        "_score": null,
        "_source":{..},
        "sort": [...],
        "join_hits": {
          "sensor": [
            {
              "_id": "s1",
              "_source": 
              {
                "model": "basic",
                "location": "Living room",
                "make": "Honeywell"
              }
            }
          ]
        }
      },
      ...
    ]
  }
}
```

- Queries with join hits are considered expensive and they are rejected if `search.allow_expensive_queries` is set to `false`.

- Join hits are fetched on the coordinating node after the fetch phase completes. And we have an optimization that tries to look up join hits during the fetch phase if the target documents reside on the same data node to reduce network requests.

Closes #76133

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/LookupJoinHitSearchPhase.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                new ResourceNotFoundException(
+                    "join hit for id [" + request.getMatchId() + "] on index [" + request.getIndex() + "] doesn't exist"));
+        }
+    }
+
+    /**
+     * An optimization of {@link org.elasticsearch.action.search.LookupJoinHitSearchPhase} that tries to lookup join hits from shards
+     * that are allocated on this node to avoid network requests if these lookup requests are executed on the coordinating node.
+     */
+    public static void tryLookupJoinHitsLocally(SearchContext searchContext, SearchHits searchHits) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/search/LookupJoinHitSearchPhase.java:206]: shouldn't the same be possible by just specifying "preference=_local" on the get request?

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/LookupJoinHitSearchPhase.java:206]: Great point. We can achieve the same thing with `preference=_only_local`. I am on it.

**Comment by dnhatn** [comment]
I am open another PR for this.

**Comment by dnhatn** [comment]
@rockdaboot Here you go: https://github.com/elastic/elasticsearch/pull/82385.

**Pull request closed without merging**